### PR TITLE
Update docs for assert! and debug_assert!

### PR DESCRIPTION
**Pull request by munyari**
Refer to #34455

### CODE REVIEW

File: src/libcore/macros.rs [hunk 1]
```diff
@@ -32,8 +32,19 @@ macro_rules! panic {
 
 /// Ensure that a boolean expression is `true` at runtime.
 ///
-/// This will invoke the `panic!` macro if the provided expression cannot be
-/// evaluated to `true` at runtime.
+/// This will ensure the termination of the program if the provided expression
+/// cannot be evaluated to `true` at runtime by means of an unrecoverable error
+/// (not necessarily a `panic!`, can also be an `abort`).
```
Comment (steveklabnik) [inline src/libcore/macros.rs]: So, here's my one concern with this wording: it _will_ always panic, the key is that panic can be either unwinding or abort. I would just stick to saying that this panics.

File: src/libcore/macros.rs [hunk 2]
```diff
@@ -32,8 +32,19 @@ macro_rules! panic {
...
+///
+/// Assertions are always checked in both debug and release builds, and cannot
+/// be disabled.
```
Comment (steveklabnik) [inline src/libcore/macros.rs]: It would be nice to point to `assert_debug` directly, if you want this behavior.

File: src/libcore/macros.rs [hunk 3]
```diff
@@ -123,6 +134,13 @@ macro_rules! assert_eq {
 /// expensive to be present in a release build but may be helpful during
 /// development.
 ///
+/// An unchecked assertion allows a program in an inconsistent state to keep
+/// running, which might have unexpected consequences but does not introduce
+/// unsafety as long as this only happens in safe code. The performance cost
+/// of assertions, is however, not measurable in general. Replacing `assert!`
+/// with `debug_assert!` is thus only encourage after thorough profiling, and
```
Comment (steveklabnik) [inline src/libcore/macros.rs]: "encouraged"

**Comment by steveklabnik** [comment]
Thanks so much! I've left a few comments; this is a solid start, just some minor tweaks :smile:

**Comment by munyari** [comment]
Thanks @steveklabnik, I've responded to each of your comments

**Comment by steveklabnik** [comment]
Looks fantastic, thank you so much!

@bors: r+ rollup

**Comment by munyari** [comment]
My first merged PR to rust! :tada:

**Pull request merged**